### PR TITLE
Deploy lambda function via s3

### LINKED ISSUE #248: Option to upload zips to S3 to bypass size limit & multi-part uploads
I'm trying to deploy a package that has a size of 109 MB (so, more than 50MB Lambda function deployment package size but less than 250MB Size of code/dependencies that you can zip into a deployment package, ref: http://docs.aws.amazon.com/lambda/latest/dg/limits.html).
To get rid of this 50MB deployment package size limitation we could do the following steps **if zip file is larger than 50 MB** (and less than 250 MB, of course) :
1. create a S3 bucket (if not exists).
2. use [putObject](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.PutObject) to upload zip file to the bucket.
3. Create/Update lambda function using `S3Bucket` and `S3Key` fields of `FunctionCode`/`UpdateFunctionCodeInput` structs (in replacement of `ZipFile` field).

What do you think about this solution?
May I propose a PR?

### PR DESCRIPTION
**Pull request by franciscocpg**
Closes #248 

## How did I implemented it?
1. Try to get the s3 bucket associated with the stack. If s3 bucket doesn't exist, create a stack with a template that adds a s3 resource and set its name to `config`. Here I am not setting a bucket name so AWS generates a random name for us. 
If there is a s3 bucket associated with this stack, set its name to `config`.
2. Create/update the lambda function. In both cases I'm using ~`s3.putObject`~ `s3manager.upload` to upload the zip code and then deploying the function using bucket/key parameters.
3. If lambda function was created (not updated) then update the stack with lambda function parameters.
4. When deleting the stack we need to empty bucket first because cloud formation fails to remove non empty buckets.

### CODE REVIEW

**Comment by tj** [comment]
wooottt thanks man! I'm pushing v0.2.0 soon, so I'll review this for v0.3.0

**Comment by Kikobeats** [comment]
awesome PR!

Probably conflicts because v2.x conflicts is on master right now.

🙏

**Comment by franciscocpg** [comment]
Hi @tj 
I've resolved the conflicts.
Any progress on this?

**Comment by tj** [comment]
I'm a little hesitant to add s3 manually for now, since Lambda uses S3 behind the scenes, I wouldn't be surprised if they just increase the size limit. The concurrent multipart-upload stuff would still be super cool though if we got that working

**Comment by franciscocpg** [comment]
got it @tj 
I'll try to replace the [s3.S3.PutObject](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.PutObject) implementation by [s3manager.Uploader.Upload](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#Uploader.Upload) (which uses multipart upload) when I have some free time.
Apparently it looks simple to do that.

**Comment by tj** [comment]
awesome! I'll try and do some testing when I'm back home, I'm curious if the concurrent uploads would really improve the upload time since upload speeds are usually ISP-bound anyway, but I suppose for CI-based deploys it could be a nice win!

**Comment by agurtovoy** [comment]
FWIW, we are deploying from CI and would appreciate this.

**Comment by kaihendry** [comment]
@agurtovoy why do you need this feature for CI? I deploy like:

```
sudo: required
language: go
go:
- tip
before_install:
- curl -sfL https://raw.githubusercontent.com/apex/up/master/install.sh | sudo sh
install:
- go get github.com/golang/dep/cmd/dep
- dep ensure -v
- dep status
script:
- cd $TRAVIS_BUILD_DIR
- go test -v ./...
after_success:
- test "$TRAVIS_BRANCH" = master && up deploy production
- up url production
```

**Comment by agurtovoy** [comment]
@kaihendry We'd appreciate faster deployment times and higher size limit (not pushing it yet, but likely to be soon enough).

**Comment by tj** [comment]
Laptop (35mb file):

- regular: ~10s
- multipart: ~11s

Laptop (512mb file):

- regular: ~3m
- multipart: ~1.5m

EC2 (512mb file):

- regular: ~30s
- multipart: ~15s

Seems worth it, especially in the cloud with a good connection. Plus that s3manager is a drop-in replacement for `s3.New()` which is nice.

For what I'd consider a "normal" sized app it seems negligible, none of my Go apps are larger than 30mb, but Node stuff gets pretty huge so it should help quite a bit there

File: config.go [hunk 1]
```diff
@@ -86,6 +86,9 @@ type Config struct {
 
 	// DNS config.
 	DNS config.DNS `json:"dns"`
+
+	// S3 Bucket Name
+	S3BucketName string
```
Comment (tj) [inline config.go]: Wonder if we should have provider specific config for stuff like this, since the bucket is only used for deploys to Lambda/AWS. Or we could maybe put it in `.lambda`'s object.

Also needs a `json` tag.

File: platform/lambda/lambda.go [hunk 2]
```diff
@@ -227,8 +222,28 @@ func (p *Platform) CreateStack(region, version string) error {
...
 }
 
+// UpdateStack implementation.
+func (p *Platform) UpdateStack(region, version string) error {
+	return stack.New(p.config, p.events, region).Update(version)
+}
+
+// SetS3BucketName based on stack.
+func (p *Platform) SetS3BucketName(region string) error {
+	return stack.New(p.config, p.events, region).SetS3BucketName()
```
Comment (tj) [inline platform/lambda/lambda.go]: How come we need this stack method? Maybe we should generate some random bucket name like `<app-name>-<uuid>` since it's just for Lambda uploads, the user likely doesn't even care to know that we happen to use S3.

That should simplify the code below. I suppose we'd have to store it somewhere on initial creation though, hmmm...

File: platform/lambda/lambda.go [hunk 3]
```diff
@@ -654,6 +747,10 @@ func (p *Platform) removeProxy() error {
 	return nil
 }
 
+func (p *Platform) getS3Key(stage string) string {
+	return fmt.Sprint(p.config.Name, "-", stage, ".zip")
```
Comment (tj) [inline platform/lambda/lambda.go]: I think we'll need some randomness here, otherwise concurrent deploys from teammates will start overriding each other.

File: platform/lambda/stack/stack.go [hunk 4]
```diff
@@ -67,6 +67,52 @@ func (s *Stack) Create(version string) error {
...
+	if err := s.report(resourceStateFromTemplate(tmpl, CreateComplete)); err != nil {
+		return errors.Wrap(err, "reporting")
+	}
+
+	stack, err := s.getStack()
+	if err != nil {
+		return errors.Wrap(err, "fetching stack")
+	}
+
+	c.S3BucketName = *stack.Outputs[0].OutputValue
```
Comment (tj) [inline platform/lambda/stack/stack.go]: are these changes necessary? not sure if it's maybe just a merge issue

**Comment by agurtovoy** [comment]
Ha, looks like I jinxed myself. Just ran into the 50Mb deployment limit, looking forward to this.

File: platform/lambda/lambda.go (hunk 2 above)
Comment (franciscocpg) [reply platform/lambda/lambda.go]: If the `up` stack is being create for the first time this PR delegates to aws cloudformation the s3 bucket name generation by creating a new cloudformation stack with just a s3 resource (using a `Logical ID`). Then it gets the generated s3 bucket name (`Physical ID`) to upload the function using s3manager and finally updates the stack with the another resources.

If the stack already exists, just sets the s3 bucket name into `config.S3BucketName` by getting its name from `DescribeStackResource`.

That said, maybe we could simplify those steps by generating the s3 bucket in `up` instead of delegating it to aws cloudformation, is that you are saying @tj?

Comment (tj) [reply platform/lambda/lambda.go]: hmm yeah I need to think on that a bit, both cases kind of suck I guess if we're introducing more latency/requests there.

Comment (franciscocpg) [reply platform/lambda/lambda.go]: Yeah, I'm not sure we need to delegate the s3 bucket name to aws cloudformation. Maybe something like `up-deploy-${service-name}` is enough.
This will save a lot of requests.

Comment (tj) [reply platform/lambda/lambda.go]: agreed! that would definitely help keep things cleaner

Comment (franciscocpg) [reply platform/lambda/lambda.go]: Ok, I'm going to implement this way then

**Comment by franciscocpg** [comment]
Hi @tj 
As we've discussed I've updated the PR with a simpler implementation for s3 bucket.
Could you please review?

**Comment by franciscocpg** [comment]
Hi @tj 
Any news?

**Comment by tj** [comment]
hey sorry! been focusing on some other stuff before adding new features. Should be done this stuff in a day or two

**Comment by franciscocpg** [comment]
ok @tj 
Just let me know when you are ready to review this PR so I can resolve the conflicting files.

**Comment by tj** [comment]
Sounds good thanks man! I think we should be ok to revisit this now. I wanted to see if reInvent would mention increased limits, but nope! So S3 is the way to go I guess.

**Comment by franciscocpg** [comment]
@tj 
Conflicts were resolved.

File: platform/lambda/lambda.go [hunk 5]
```diff
@@ -393,12 +382,12 @@ func (p *Platform) deploy(region, stage string) (version string, err error) {
 
 	defer func() {
 		fields["duration"] = time.Since(start)
-		fields["version"] = version
```
Comment (tj) [inline platform/lambda/lambda.go]: need to keep this part, I'll tweak it back to the old semantics, not seeing a reason why we need the CreateStack within deploy()

**Comment by tj** [comment]
we also need to account for existing projects which may not have the bucket, but with Fargate maybe it's not even worth trying to hack this stuff into Lambda, I think I'd rather prioritize Fargate, we need S3 cleanup as well

File: platform/lambda/lambda.go (hunk 5 above)
Comment (franciscocpg) [reply platform/lambda/lambda.go]: You're right about that. I'm changing back to the old semantics.

**Comment by franciscocpg** [comment]
Hi @tj 
I've made the changes you mentioned:
- Rollback `deploy()` code to the old semantics.
- Account for existing projects which may not have the bucket

Could you review again?

**Comment by tj** [comment]
awesome thanks man! looks great. It's at least ~35% faster for a 28mb app, curious how it'll impact even larger apps.

the only thing we need to figure out now is the S3 bucket name, `up-deploy-NAME` will clash with others since S3 buckets are globally scoped at the region level.

The only thing I can think of is fetching the account ID, or we could use a random name and save it to up.json so it can be referenced later but that feels more hacky.

**Comment by hhsnopek** [comment]
Curious, what if all up deploys went into `up-deploy/` and each "deploy" followed the naming convention of `appName/region/version`? then at the end of it all you'd never clash with names and you'd have all deploys in one s3 bucket

example path: `up-deploys/gh-polls/us-east-1/xxx`

**Comment by tj** [comment]
ah can grab the account id from the role we prime, so that'll be ok.


@hhsnopek yea one bucket might not be bad, but complete isolation might prevent some accidental deletions

**Comment by lukeed** [comment]
Sorry if this is a stupid question, I'm sure it is, but is S3 a valuable target once Fargate is available? Is there anyway we can save a lot of this hard work for Fargate compat?

**Comment by tj** [comment]
@lukeed yes and no I guess. It's definitely less important with Fargate as an option, but S3's multi-part deploys speed things up for the Lambda case.

Lifting the size restriction isn't a huge win IMO, since an app that huge is doing something wrong in my opinion haha, but faster uploads is nice.

**Comment by franciscocpg** [comment]
:v:
